ncm-spma/src/main/perl/spma/apt.pm

```diff
@@ -0,0 +1,433 @@
+            }
+        }
+        my $rs = $self->template()->process($template, $repo, $fh);
+        if (!$rs) {
+            $self->error ("Unable to generate repository $repo->{name}: ", $self->template()->error());
+            $fh->cancel();
+            return 0;
+        }
+        $changes += $fh->close() || 0; # handle undef
+        $fh->close();
+    }
+
+    return $changes;
+}
+
+
+# Returns a set of all installed packages
+sub get_installed_pkgs
+{
+    my $self = shift;
+    $self->debug(5, 'Entered installed_pkgs()');
+
+    my $cmd = CAF::Process->new($CMD_DPKG_QUERY, keeps_state => 1, log => $self);
+
+    my $out = $cmd->output();
+    if ($?) {
+        return 0;
+    }
+    my @pkgs = grep(m{^ii }, split(/\n/, $out));
+    for (@pkgs) {s/^...;//g}; # Strip package status characters
+
+    return Set::Scalar->new(@pkgs);
+}
+
+
+sub get_package_version_arch
+{
+    my ($self, $name, $details) = @_;
+    $self->debug(5, 'Entered get_package_version_arch()');
+
+    my @versions;
+
+    if ($details) {
+        while (my ($ver, $archs) = each(%$details)) {
+            if ($archs->{arch}) {
+                while (my $arch = each(%$archs->{arch})) {
+                    $self->debug(5, '  Adding package ', $name, ' with version ', $ver, ' and architecture ', $arch, ' to list');
+                    push(@versions, sprintf('%s:%s=%s', $name, $arch, unescape($ver)));
+                }
+            }
+            else {
+                $self->debug(5, '  Adding package ', $name, ' with version ', $ver, ' but without architecture to list');
+                push(@versions, sprintf('%s=%s', $name, unescape($ver)));
+            }
+        }
+    }
+    else {
+        $self->debug(5, '  Adding package ', $name, ' without version or architecture to list');
+        push(@versions, $name);
+    }
+
+    return \@versions;
+}
+
+
+sub apply_package_version_arch
+{
+    my ($self, $packagelist, $packagetree) = @_;
+
+    $self->debug(5, 'Entered apply_package_version_arch()');
+
+    my @packages = @$packagelist;
+    my @results;
+
+    while (my $name = shift(@packages)) {
+        my $name_escaped = escape($name);
+        if (exists($packagetree->{$name_escaped})) {
+            my $versions = $self->get_package_version_arch($name, $packagetree->{$name_escaped});
+            push(@results, @$versions);
+        }
+        else {
+            $self->error('Could not find package ', $name, ' in tree');
+        }
+    }
+
+    return \@results;
+}
+
+
+# Return a set of desired packages.
+sub get_desired_pkgs
+{
+    my ($self, $pkgs) = @_;
+    $self->debug(5, 'Entered get_desired_pkgs()');
+
+    my $packages = Set::Scalar->new();
+
+    while (my $name_escaped = each(%$pkgs)) {
+        my $name = unescape($name_escaped);
+        if (!$name) {
+            $self->error("Invalid package name: ", unescape($name));
+            return 0;
```

Comment:
return in the middle of an each... :see_no_evil: :hear_no_evil: :speak_no_evil:  https://github.com/quattor/CAF/issues/109